dll/win32/shell32/shellrecyclebin/recyclebin_v5_enumerator.cpp

```diff
@@ -0,0 +1,462 @@
+    return hr;
+}
+
+STDMETHODIMP RecycleBin5File::GetPhysicalFileSize(ULARGE_INTEGER *pPhysicalFileSize)
+{
+    TRACE("(%p, %p)\n", this, pPhysicalFileSize);
+    pPhysicalFileSize->u.HighPart = 0;
+    pPhysicalFileSize->u.LowPart = m_deletedFile.dwPhysicalFileSize;
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::GetAttributes(DWORD *pAttributes)
+{
+    DWORD dwAttributes;
+
+    TRACE("(%p, %p)\n", this, pAttributes);
+
+    dwAttributes = GetFileAttributesW(m_FullName);
+    if (dwAttributes == INVALID_FILE_ATTRIBUTES)
+        return HRESULT_FROM_WIN32(GetLastError());
+
+    *pAttributes = dwAttributes;
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::GetFileName(SIZE_T BufferSize, LPWSTR Buffer, SIZE_T *RequiredSize)
+{
+    DWORD dwRequired;
+
+    TRACE("(%p, %u, %p, %p)\n", this, BufferSize, Buffer, RequiredSize);
+
+    dwRequired = (DWORD)(wcslen(m_deletedFile.FileNameW) + 1) * sizeof(WCHAR);
+    if (RequiredSize)
+        *RequiredSize = dwRequired;
+
+    if (BufferSize == 0 && !Buffer)
+        return S_OK;
+
+    if (BufferSize < dwRequired)
+        return E_OUTOFMEMORY;
+    CopyMemory(Buffer, m_deletedFile.FileNameW, dwRequired);
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::GetTypeName(SIZE_T BufferSize, LPWSTR Buffer, SIZE_T *RequiredSize)
+{
+    HRESULT hr;
+    DWORD dwRequired;
+    DWORD dwAttributes;
+    SHFILEINFOW shFileInfo;
+
+    TRACE("(%p, %u, %p, %p)\n", this, BufferSize, Buffer, RequiredSize);
+
+    hr = GetAttributes(&dwAttributes);
+    if (!SUCCEEDED(hr))
+        return hr;
+
+    hr = SHGetFileInfoW(m_FullName, dwAttributes, &shFileInfo, sizeof(shFileInfo), SHGFI_TYPENAME | SHGFI_USEFILEATTRIBUTES);
+    if (!SUCCEEDED(hr))
+        return hr;
+
+    dwRequired = (DWORD)(wcslen(shFileInfo.szTypeName) + 1) * sizeof(WCHAR);
+    if (RequiredSize)
+        *RequiredSize = dwRequired;
+
+    if (BufferSize == 0 && !Buffer)
+        return S_OK;
+
+    if (BufferSize < dwRequired)
+        return E_OUTOFMEMORY;
+    CopyMemory(Buffer, shFileInfo.szTypeName, dwRequired);
+    return S_OK;
+}
+
+STDMETHODIMP RecycleBin5File::Delete()
+{
+    TRACE("(%p)\n", this);
+    return m_recycleBin->Delete(m_FullName, &m_deletedFile);
+}
+
+STDMETHODIMP RecycleBin5File::Restore()
+{
+    TRACE("(%p)\n", this);
+    return m_recycleBin->Restore(m_FullName, &m_deletedFile);
+}
+
+RecycleBin5File::RecycleBin5File(
+    IN IRecycleBin5 *prb,
+    IN LPCWSTR Folder,
+    IN PDELETED_FILE_RECORD pDeletedFile,
+    IN OUT LPWSTR pszFullName)
+    : m_ref(1)
+    , m_recycleBin(prb)
+    , m_deletedFile(*pDeletedFile)
+    , m_FullName(pszFullName)
+{
+    m_recycleBin->AddRef();
+}
+
+static HRESULT
+RecycleBin5File_Constructor(
+    IN IRecycleBin5 *prb,
+    IN LPCWSTR Folder,
+    IN PDELETED_FILE_RECORD pDeletedFile,
+    OUT IRecycleBinFile **ppFile)
+{
+    if (!ppFile)
+        return E_POINTER;
+
+    LPCWSTR Extension = wcsrchr(pDeletedFile->FileNameW, '.');
+    WCHAR szFullName[MAX_PATH];
```

Comment:
I'm not fond of this new local buffer, while the original code was able to avoid using one.
Original code was calculating the necessary size of the string, then allocated a buffer of proper size before doing the wsprintf call.
(Also consider using a stringsafe function here?)

In addition, the old code was doing some extra checks on the retrieved "Extension" pointer. Wouldn't they be needed there as well? Or, what about using `PathFindExtensionW()` instead, if that latter function also does some extra "validation" on where the extension must be found?

Reply:
Fixed by using ATL `CStringW` and `PathFindExtensionW`.